### DISCUSSION TITLE
Form Requests in Livewire

### DISCUSSION THREAD

**Discussion by ju5t** [livewire/livewire#5171] [2. Show and tell]
Our application has a lot of full page Livewire components. Nearly all of them have forms and quite a few of them have to do a lot of validation when a form is submitted. I have always found this a bit messy. It adds a lot of code just for validation. And then there's all the custom validation messages. It quickly feels bloated.

The concept of Form Requests is pretty familiar with a lot of Laravel developers. And although our implementation may be a bit rough around the edges and is not 100% the same. You can find our solution as a trait named [HasFormRequest](https://gist.github.com/ju5t/b5046ebb69ab5212620f1822dfbee40a).

The gist of it is:

```php
class UserEdit extends Component
{
    use HasFormRequest;
  
    /**
     * @var class-string<FormRequest>
     */
    protected string $request = UserEditRequest::class;
}
```

You can define `rules()` and `messages()` in the form request, being `UserEditRequest` in this example. If you specify `authorize()` it will call it automatically, so you can move your authorization logic in your request too.

We found it very useful in our app. Hope you all find it useful too!

**Comment by joshhanley** [comment]
@ju5t nice! Thanks for sharing 😁

**Comment by JuliusMaskow** [comment]
I had the same issue in larger Livewire components and ended up releasing a small package that combines Laravel Form Requests with Livewire components in a clean way. Maybe it helps someone:

https://github.com/MaskowSoftwares/livewire-combined-request
